wal/decoder.go

```diff
@@ -28,30 +28,67 @@ import (
 )
 
 type decoder struct {
-	mu sync.Mutex
-	br *bufio.Reader
+	mu  sync.Mutex
+	brs []*bufio.Reader
 
-	crc hash.Hash32
+	lastValidOff int64
+	crc          hash.Hash32
 }
 
-func newDecoder(r io.Reader) *decoder {
+func newDecoder(r ...io.Reader) *decoder {
+	readers := make([]*bufio.Reader, len(r))
+	for i := range r {
+		readers[i] = bufio.NewReader(r[i])
+	}
 	return &decoder{
-		br:  bufio.NewReader(r),
+		brs: readers,
 		crc: crc.New(0, crcTable),
 	}
 }
 
 func (d *decoder) decode(rec *walpb.Record) error {
+	rec.Reset()
+
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
-	rec.Reset()
-	l, err := readInt64(d.br)
+	err := d.decodeRecord(rec)
+	if err != walpb.ErrCRCMismatch {
```

Comment:
add a comment for this? // ErrCRCMismatch might be caused by ... so we need to continue checking.